src/Components/AdvancedSearch/Internal/QueryBuilder/QueryBuilder.tsx

```diff
@@ -28,37 +29,88 @@ const QueryBuilder: React.FC<IQueryBuilderProps> = (props) => {
         allowedPropertyValueTypes,
         executeQuery,
         styles,
+        theme,
         updateColumns
     } = props;
-    const classNames = getClassNames(styles, {
-        theme: useTheme()
-    });
-    // This could be a single map???
+
+    // State
     const querySnippets = useRef(new Map<string, QueryRowType>());
     const validityMap = useRef(new Map<string, boolean>());
     const propertyNames = useRef(new Map<string, string>());
-    //
     const [isSearchDisabled, setIsSearchDisabled] = useState(true);
     const [rows, updateRows] = useState<any[]>([
         {
             rowId: String(Math.random())
         }
     ]);
 
-    const checkIsValidQuery = () => {
-        if (validityMap.current.size === 0) {
-            return true;
-        }
-        let isValid = true;
-        validityMap.current.forEach((validity: boolean) => {
-            if (!validity) {
-                isValid = false;
+    // Hooks
+    const { t } = useTranslation();
+
+    // Classname after state to track row #
+    const classNames = getClassNames(styles, {
+        theme: useTheme(),
+        rowCount: rows.length
+    });
+
+    // Callbacks
+    const removeRow = useCallback(
+        (index: number, rowId: string) => {
+            // Remove row component
+            const newRows = [...rows];
+            if (index === 0) {
+                newRows.shift();
+            } else {
+                newRows.splice(index, 1);
             }
-        });
-        return isValid;
+            updateRows(newRows);
+            // Remove row values in query and property
+            querySnippets.current.delete(rowId);
+            propertyNames.current.delete(rowId);
+            validityMap.current.delete(rowId);
+            // Reset validity
+            setIsSearchDisabled(!checkIsValidQuery());
+        },
+        [rows]
+    );
+
+    const onChangeValue = useCallback((rowId: string, newValue: string) => {
+        // Reset validity on value change
+        if (newValue.length) {
+            validityMap.current.set(rowId, true);
+        } else {
+            validityMap.current.set(rowId, false);
+        }
+        setIsSearchDisabled(!checkIsValidQuery());
+    }, []);
```

Comment:
if you are going to `useCallback` on these, make sure your dependency lists are updated. I'd suggest just not using `useCallback` here unless you suspect that it will break a memoization lower in the DOM

Reply:
using them to avoid re-renders of input fields. Should not be a huge performance upgrade but at least it should save us some renders.